src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -518,5 +577,88 @@ internal JsonArrayPool(ArrayPool<T> arrayPool)
 
             public void Return(T[] array) => this.arrayPool.Return(array);
         }
+
+        private class JsonProgressClientConverter : JsonConverter
+        {
+            private JsonMessageFormatter formatter = null;
+
+            public JsonProgressClientConverter(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter;
+            }
+
+            public override bool CanConvert(Type objectType) => MessageFormatterHelper.FindIProgressOfT(objectType) != null;
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                // if the requestId is empty it means the Progress object comes from a response
+                if (this.formatter.requestId == null)
+                {
+                    writer.WriteValue(value);
+                    return;
```

Comment:
I'm not sure what this does. Does it result in us being recursively called back since `value` is an `IProgress<T>` instance that otherwise can't be serialized? I suspect the better option here would be to throw `NotSupportedException` with a message explaining why it doesn't work.